### PR TITLE
Support restarting OmegaT

### PR DESCRIPTION
**Pull request by miurahr**
Implement [RFE#1564](https://sourceforge.net/p/omegat/feature-requests/1564/)

- Introduce Main.restartGUI() static method to support graceful restart.
- Add Project-Restart menu item

This is also a preparation for #69

Signed-off-by: Hiroshi Miura <miurahr@linux.com>

### CODE REVIEW

File: src/org/omegat/gui/main/MainWindowMenu.java [hunk 1]
```diff
@@ -291,11 +291,13 @@ public void menuCanceled(MenuEvent e) {
...
         });
         projectExitMenuItem = createMenuItem("TF_MENU_FILE_QUIT");
+        projectRestartMenuItem = createMenuItem("TF_MENU_FILE_RESTART");
 
         // all except MacOSX
         if (!Platform.isMacOSX()) {
             projectMenu.addSeparator();
             projectMenu.add(projectExitMenuItem);
+            projectMenu.add(projectRestartMenuItem);
         }
```
Comment (amake) [inline src/org/omegat/gui/main/MainWindowMenu.java:301]: The menu item doesn't appear to be available anywhere on macOS. I think it should go *somewhere*. Maybe at the bottom of the Project menu.

Also I'm not sure about putting it below Exit; I think it's probably strongly expected that Exit is at the very bottom. What about Restart and then Exit?

So macOS would have

- (other Project menu content)
- divider
- Restart

Other platforms would have the above *and* then:

- Exit

```suggestion
        projectMenu.add(projectRestartMenuItem);
        projectMenu.addSeparator();

        // all except MacOSX
        if (!Platform.isMacOSX()) {
            projectMenu.add(projectExitMenuItem);
        }
```

File: src/org/omegat/gui/main/MainWindowMenuHandler.java [hunk 2]
```diff
@@ -368,6 +368,63 @@ private void openFile(File path) {
         }
     }
 
+    /** Restart OmegaT */
+    public void projectRestartMenuItemActionPerformed() {
```
[the remark this answers is not in the document]
Comment (miurahr) [reply src/org/omegat/gui/main/MainWindowMenuHandler.java:372]: done.

File: src/org/omegat/Main.java [hunk 3]
```diff
@@ -208,6 +216,36 @@ public static void main(String[] args) {
         }
     }
 
+    public static void restartGUI(String projectDir) {
+        Log.log("===============================================================");
+        Log.log("===              Restart OmegaT                            ====");
+        Log.log("===============================================================");
+        final String javaBin = System.getProperty("java.home") + File.separator + "bin" + File.separator + "java";
+        try {
+            final File currentJar;
```
[the remark this answers is not in the document]
Comment (miurahr) [reply src/org/omegat/Main.java:225]: removed.

File: src/org/omegat/Main.java [hunk 4]
```diff
@@ -208,6 +216,36 @@ public static void main(String[] args) {
         }
     }
 
+    public static void restartGUI(String projectDir) {
+        Log.log("===============================================================");
+        Log.log("===              Restart OmegaT                            ====");
+        Log.log("===============================================================");
```
[the remark this answers is not in the document]
Comment (miurahr) [reply src/org/omegat/Main.java:222]: reduced banner size.

File: src/org/omegat/Main.java [hunk 5]
```diff
@@ -208,6 +216,36 @@ public static void main(String[] args) {
...
+            command.add(javaBin);
+            command.addAll(JVMPARAMS);
+            command.add("-cp");
+            command.add(ManagementFactory.getRuntimeMXBean().getClassPath());
+            command.add(Main.class.getName());
+            PARAMS.forEach((k, v) -> {
+                if (!k.equals(CLIParameters.PROJECT_DIR)) {
+                    command.add(k + "=" + v);
+                }
+            });
```
[the remark this answers is not in the document]
Comment (miurahr) [reply src/org/omegat/Main.java:237]: I forget add "--" to argument.

File: src/org/omegat/Main.java [hunk 6]
```diff
@@ -122,6 +128,10 @@ public static void main(String[] args) {
             TeamTool.main(Arrays.copyOfRange(args, 1, args.length));
         }
 
+        // retrieve JVM parameters.
+        RuntimeMXBean runtimeMxBean = ManagementFactory.getRuntimeMXBean();
+        JVMPARAMS.addAll(runtimeMxBean.getInputArguments());
```
Comment (amake) [inline src/org/omegat/Main.java:133]: Is there a reason we need to store these at startup? It seems that it's probably only used at restart time, so we might as well just wait until then.

**Review by amake — COMMENTED** [review]
I tested your latest changes and now arguments are being passed correctly. Thanks.

Just one more small point; otherwise LGTM.

File: src/org/omegat/Main.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (amake) [reply src/org/omegat/Main.java]: Thanks, it's working now.

There is some coupling between this logic and the arg parsing logic. I made a comment in `CLIParameters.java`.

Maybe the arg-reconstructing logic should move to `CLIParameters` as well.

[the remark this answers is not in the document]
Comment (amake) [reply src/org/omegat/Main.java]: > Maybe the arg-reconstructing logic should move to `CLIParameters` as well.

Done in https://github.com/omegat-org/omegat/pull/91/commits/b0cb66e2bb10957fa0d446ec11383885a8e8bd7b

File: src/org/omegat/Main.java (hunk 6 above)
Comment (miurahr) [reply src/org/omegat/Main.java:133]: Yes, it is only used at restart time.

**Comment by amake** [comment]
Thank you! I've squashed to `master`.

**Pull request closed without merging**